### PR TITLE
pubsub: wrap errors with gcerrors

### PR DESCRIPTION
**Pull request by jba**
Wrap driver errors using the gcerrors package.

### CODE REVIEW

File: pubsub/pub_test.go [hunk 1]
```diff
@@ -19,8 +20,6 @@ func (t *funcTopic) SendBatch(ctx context.Context, ms []*driver.Message) error {
 
 func (s *funcTopic) IsRetryable(error) bool { return false }
```
Comment (vangent) [inline pubsub/pub_test.go]: Nit: could be deleted too, no?

File: pubsub/pubsub.go [hunk 2]
```diff
@@ -369,3 +372,22 @@ func defaultAckBatcher(ctx context.Context, s *Subscription) driver.Batcher {
...
+
+func wrapError(ec errorCoder, err error) error {
+	if err == nil {
+		return nil
+	}
+	// Don't wrap context errors.
+	if _, ok := err.(*retry.ContextError); ok || err == context.Canceled || err == context.DeadlineExceeded {
+		return err
+	}
+	// Don't double-wrap.
```
Comment (vangent) [inline pubsub/pubsub.go]: When would this happen? It seems like it would just be hiding a bug.

File: pubsub/pubsub.go [hunk 3]
```diff
@@ -369,3 +372,22 @@ func defaultAckBatcher(ctx context.Context, s *Subscription) driver.Batcher {
 	}
 	return batcher.New(reflect.TypeOf([]driver.AckID{}).Elem(), maxHandlers, handler)
 }
+
+type errorCoder interface {
+	ErrorCode(error) gcerrors.ErrorCode
+}
+
+func wrapError(ec errorCoder, err error) error {
+	if err == nil {
```
Comment (vangent) [inline pubsub/pubsub.go]: When would this happen? Seems like it's hiding a bug.

File: pubsub/sub_test.go [hunk 4]
```diff
@@ -44,8 +45,6 @@ func (s *scriptedSub) SendAcks(ctx context.Context, ackIDs []driver.AckID) error
 
 func (s *scriptedSub) IsRetryable(error) bool { return false }
```
Comment (vangent) [inline pubsub/sub_test.go]: Remove?

File: pubsub/pub_test.go (hunk 1 above)
Comment (jba) [reply pubsub/pub_test.go]: Done.

**Comment by jba** [comment]
PTAL.
I still need to do `ErrorAs`, but I'll do that in a separate PR if that's OK.

File: internal/gcerr/gcerr.go [hunk 5]
```diff
@@ -50,6 +50,9 @@ const (
 
 	// The feature is not implemented.
 	Unimplemented ErrorCode = 6
+
+	// The system was in the wrong state.
+	FailedPrecondition ErrorCode = 7
```
Comment (vangent) [inline internal/gcerr/gcerr.go]: Doesn't `gcerr_string.go` need to be updated for this?

File: pubsub/pubsub.go (hunk 3 above)
Comment (vangent) [reply pubsub/pubsub.go]: `if err == nil` wasn't deleted?

File: internal/gcerr/gcerr.go (hunk 5 above)
Comment (jba) [reply internal/gcerr/gcerr.go]: The `stringer` tool is currently broken in module mode. I'm going to work with the tools team in New York to work around that. I'll address in a followup PR.

File: pubsub/pubsub.go (hunk 3 above)
Comment (vangent) [reply pubsub/pubsub.go]: Ping on this one.

File: internal/gcerr/gcerr.go (hunk 5 above)
Comment (vangent) [reply internal/gcerr/gcerr.go]: Add a TODO?

File: pubsub/pubsub.go (hunk 3 above)
Comment (jba) [reply pubsub/pubsub.go]: Sorry, missed it. Done now.

File: internal/gcerr/gcerr.go (hunk 5 above)
Comment (jba) [reply internal/gcerr/gcerr.go]: Done.

**Comment by jba** [comment]
I also had to add back an ErrorCode method on a test driver. It failed during the MacOS travis run. Not sure what that's about...

PTAL

**Pull request merged**